modules/nixos.nix

```diff
@@ -34,29 +34,127 @@ in
-      })
-      config.systems.nixos;
+    assertions = lib.lists.flatten (lib.attrs.mapToList
+      (name: value: let
+        hasNixpkgs = !(builtins.isNull value.pkgs);
+        requestedHomes = value.homes != [];
+        hasHomeManager = !(builtins.isNull value.home-manager);
+      in [
+        {
+          assertion = hasNixpkgs;
+          message = "A Nixpkgs instance is required for the NixOS system \"${name}\", but none was provided and \"inputs.nixpkgs\" does not exist.";
+        }
+        {
+          assertion = !requestedHomes || hasHomeManager;
+          message = "A home-manager instance is required to enable homes for the NixOS system \"${name}\", but none was provided and \"inputs.home-manager\" does not exist.";
+        }
+        (map (home: let
+          hasHome = global.config ? homes.${home};
+          homeHasHomeManager = !(builtins.isNull global.config.homes.${home}.home-manager);
+          homeIsValidForSystem = global.config.homes.${home} ? result.${value.pkgs.system};
+        in [
+          {
+            assertion = hasHome;
+            message = "You've asked for the home \"${home}\" to be activated in the NixOS system \"${name}\", but it doesn't exist. Please set it up using [the Nilla Home plugin](https://github.com/nilla-nix/home).";
+          }
+          {
+            assertion = !hasHome || homeHasHomeManager;
+            message = "You've asked for the home \"${home}\" to be activated in the NixOS system \"${name}\", but it needs a home-manager instance, none was provided and \"inputs.home-manager\" does not exist.";
+          }
+          {
+            assertion = !hasHome || !homeHasHomeManager || !hasNixpkgs || homeIsValidForSystem;
+            message = "You've asked for the home \"${home}\" to be activated in the NixOS system \"${name}\", but it isn't valid for \"${value.pkgs.system}\" systems.";
+          }
+        ]) value.homes)
+        (let
+          usernames = map (home: let
+            hasHome = global.config ? homes.${home};
+            homeHasHomeManager = !(builtins.isNull global.config.homes.${home}.home-manager);
+            homeIsValidForSystem = global.config.homes.${home} ? result.${value.pkgs.system};
+          in 
+            if hasHome && homeHasHomeManager && hasNixpkgs && homeIsValidForSystem
+            then let
+              homeNameParts = builtins.match "([a-z][-a-z0-9]*)(@([-A-Za-z0-9]+))?(:([-_A-Za-z0-9]+))?" home;
+              username = builtins.elemAt homeNameParts 0;
+            in username
+            else null) value.homes;
+          existingUsernames = builtins.filter (username: username != null) usernames;
+          uniqueUsernames = lib.lists.unique existingUsernames;
+        in {
+          assertion = !hasNixpkgs || (existingUsernames == uniqueUsernames);
+          message = "There are multiple homes for a single user in the NixOS system \"${name}\". Please make sure you've only enabled a single home per user.";
```

Comment:
I feel like this should be possible since the configurations can be merged.

Reply:
maybe, but it does make args more fiddly and there is no problem with *expanding* the permitted contract later - it's contracting it that'll cause issues